File: src/test/java/no/nav/dokgen/services/TemplateServiceTests.java
package no.nav.dokgen.services;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Java6Assertions.catchThrowable;

import java.io.FileWriter;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import org.apache.commons.io.FileUtils;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.github.jknack.handlebars.Template;

import no.nav.dokgen.resources.TemplateResource;
import no.nav.dokgen.util.FileStructureUtil;

public class TemplateServiceTests {
    private static final String MALNAVN = "lagretMal";
    private static final String MARKDOWN = "# Hei, {{name}}";
    private static final String MARKDOWN_CONTENT = "\"" + MARKDOWN + "\"";
    private static final String MERGE_FIELDS = "{\"name\": \"Peter\"}";
    private static final String GYLDIG_PAYLOAD = lagTestPayload(MARKDOWN_CONTENT, MERGE_FIELDS);
    private static final String FOOTER_CONTENT = "this would be the footer.";

    private static String lagTestPayload(String markdownContent, String interleavingFields) {
        return "{\"markdownContent\": " + markdownContent +
                ", \"interleavingFields\": " + interleavingFields +
                ", \"useTestSet\": false}";
    }

    public static Path getTestContentPath() {
        try {
            return Paths.get(
                    DocumentGeneratorServiceTests.class.getProtectionDomain().getCodeSource().getLocation().toURI()
            ).resolve(
                    Paths.get("test-content")
            ).toAbsolutePath();
        } catch (URISyntaxException e) {
            return null;
        }
    }

    private static final String TOM_JSON = "{}";

    private TemplateService malService;
    private Path contentRoot;
    private Path templateRoot;

    @Rule
    public TemporaryFolder temporaryFolder = new TemporaryFolder();

    @Before
    public void setUp() throws IOException {
        contentRoot = temporaryFolder.getRoot().toPath();
        templateRoot = FileStructureUtil.getTemplateRootPath(contentRoot);
        Files.createDirectories(templateRoot);
        FileWriter fileWriter = new FileWriter(templateRoot.resolve("footer.hbs").toString());
        fileWriter.write("\n" + FOOTER_CONTENT);
        fileWriter.close();
        malService = new TemplateService(contentRoot, new DocumentGeneratorService(getTestContentPath()), new JsonService(contentRoot));
    }

    @Test
    public void skalHentAlleMalerReturnererTomtSett() {
        assertThat(malService.listTemplates()).isEmpty();
    }

    @Test
    public void skalHentAlleMaler() throws IOException {
        Files.createDirectory(templateRoot.resolve("MAL1"));
        assertThat(malService.listTemplates()).containsExactly("MAL1");
    }

    @Test
    public void skalReturnereExceptionVedHentingAvMal() {

        // expect
        Throwable thrown = catchThrowable(() -> malService.getTemplate("IkkeEksisterendeMal"));

        // when
        assertThat(thrown).isInstanceOf(RuntimeException.class).hasMessage("Kan ikke finne mal med navn IkkeEksisterendeMal");
    }

    @Test
    public void skalReturnereMal() throws IOException {
        Path malPath = templateRoot.resolve("MAL");
        Files.createDirectories(malPath);
        Path templatePath = FileStructureUtil.getTemplatePath(contentRoot, "MAL");
        Files.write(templatePath, "maldata".getBytes());


        // expect
        TemplateResource template = malService.getTemplate("MAL");

        // when
        assertThat(template.getContent()).isEqualTo("maldata");
    }

    @Test
    public void skalReturnereIllegalArgumentExceptionVedLagreMalUtenPayloadUtenMarkdown() {

        // expect
        Throwable thrown = catchThrowable(() -> malService.saveTemplate("tomPayload", "{}"));

        assertThat(thrown).isInstanceOf(IllegalArgumentException.class).hasMessage("Kan ikke hente markdown for payload={}");
    }

    @Test
    public void skalLagreMal() throws IOException {

        // expect
        malService.saveTemplate(MALNAVN, GYLDIG_PAYLOAD);

        String malData = Files.readString(FileStructureUtil.getTemplatePath(contentRoot, MALNAVN));
        assertThat(malData).isEqualTo(MARKDOWN);
    }


    @Test
    public void skal_compile_inline_with_a_footer() throws IOException {
        Template tmp = malService.compileInLineTemplate("this would be a header\n{{> footer }}");
        JsonNode node = JsonNodeFactory.instance.arrayNode();
        String compiledStr = tmp.apply(node);
        assertThat(compiledStr).endsWith(FOOTER_CONTENT);
    }

    @Test
    public void skal_overskrive_alt_ved_lagring_av_nytt_slankere_innhold() throws IOException {
        skalLagreMal();

        malService.saveTemplate(MALNAVN, lagTestPayload("\"" + MARKDOWN.substring(3) + "\"", MERGE_FIELDS));

        String malData = Files.readString(FileStructureUtil.getTemplatePath(contentRoot, MALNAVN));
        assertThat(malData).isEqualTo(MARKDOWN.substring(3));
    }

    @Test
    public void skalHenteHtml() throws IOException {
        String malNavn = "html";

        FileUtils.writeStringToFile(FileStructureUtil.getTemplateSchemaPath(contentRoot, malNavn).toFile(),
                TOM_JSON,
                StandardCharsets.UTF_8
        );

        // expect
        malService.saveTemplate(malNavn, GYLDIG_PAYLOAD);
        String html = malService.createHtml(malNavn, "{\"name\":\"Peter\"}");
        assertThat(html).contains("<h1>Hei, Peter</h1>");
    }

    @Test
    public void skalHentePdf() throws IOException {
        String malNavn = "pdf";

        FileUtils.writeStringToFile(FileStructureUtil.getTemplateSchemaPath(contentRoot, malNavn).toFile(),
                TOM_JSON,
                StandardCharsets.UTF_8
        );

        // expect
        malService.saveTemplate(malNavn, GYLDIG_PAYLOAD);
        byte[] pdf = malService.createPdf(malNavn, GYLDIG_PAYLOAD);
        assertThat(pdf[1]).isEqualTo((byte) 0x50);//P
        assertThat(pdf[2]).isEqualTo((byte) 0x44);//D
        assertThat(pdf[3]).isEqualTo((byte) 0x46);//F
    }

    @Test
    public void skalHentePdfFraVariation() throws IOException {
        String malNavn = "pdf";
        String templateVariation = "template_NN";

        FileUtils.writeStringToFile(FileStructureUtil.getTemplateSchemaPath(contentRoot, malNavn).toFile(),
                TOM_JSON,
                StandardCharsets.UTF_8
        );

        // expect
        malService.saveTemplate(malNavn, GYLDIG_PAYLOAD, templateVariation);
        byte[] pdf = malService.createPdf(malNavn, GYLDIG_PAYLOAD, templateVariation);
        assertThat(pdf[1]).isEqualTo((byte) 0x50);//P
        assertThat(pdf[2]).isEqualTo((byte) 0x44);//D
        assertThat(pdf[3]).isEqualTo((byte) 0x46);//F
    }
}

File: src/test/kotlin/no/nav/dokgen/services/TemplateServiceTests.kt
package no.nav.dokgen.services

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.JsonNodeFactory
import com.github.jknack.handlebars.Context
import com.github.jknack.handlebars.jackson.JsonNodeValueResolver
import no.nav.dokgen.configuration.ContentProperties
import no.nav.dokgen.util.FileStructureUtil.getTemplateRootPath
import no.nav.dokgen.util.FileStructureUtil.getTemplatePath
import no.nav.dokgen.util.FileStructureUtil.getTemplateSchemaPath
import org.apache.commons.io.FileUtils
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import org.springframework.core.io.ClassPathResource
import java.io.FileWriter
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

class TemplateServiceTests {
    private lateinit var malService: TemplateService
    private lateinit var templateRootTemp: Path

    @TempDir
    lateinit var basePathTemp: Path

    private fun testContentProps(root: Path) = ContentProperties().apply { this.root = root }

    @BeforeEach
    @Throws(IOException::class)
    fun setUp() {
        templateRootTemp = getTemplateRootPath(basePathTemp)
        Files.createDirectories(templateRootTemp)

        val staticContentForCopy: Path = ClassPathResource("/test-content").file.toPath().toAbsolutePath()
        FileUtils.copyDirectory(staticContentForCopy.toFile(), basePathTemp.toFile())

        val fileWriter = FileWriter(templateRootTemp.resolve("footer.hbs").toString())
        fileWriter.write(
            """
    
    $FOOTER_CONTENT
    """.trimIndent()
        )
        fileWriter.close()
        val props = testContentProps(basePathTemp)
        malService =
            TemplateService(
                props,
                DocumentGeneratorService(props),
                JsonService(props)
            )
    }

    @Test
    fun skalHentAlleMalerReturnererDefaultSett() {
        Assertions.assertThat(malService.listTemplates()).containsExactly("pdf", "html", "lagretMal")
    }

    @Test
    @Throws(IOException::class)
    fun skalHentAlleMaler() {
        Files.createDirectory(templateRootTemp.resolve("MAL1"))
        Assertions.assertThat(malService.listTemplates()).contains("MAL1")
    }

    @Test
    fun skalReturnereExceptionVedHentingAvMal() {
        assertThatThrownBy { malService.getTemplate("IkkeEksisterendeMal") }
            .isInstanceOf(RuntimeException::class.java)
            .hasMessage("Kan ikke finne mal med navn IkkeEksisterendeMal")
    }

    @Test
    fun skalReturnereMal() {
        val malPath = templateRootTemp.resolve("MAL")
        Files.createDirectories(malPath)
        val templatePath = getTemplatePath(basePathTemp, "MAL")
        Files.write(templatePath, "maldata".toByteArray())


        // expect
        val (_, content) = malService.getTemplate("MAL")

        // when
        Assertions.assertThat(content).isEqualTo("maldata")
    }

    @Test
    fun skalReturnereMalDerMalnavnErEnPath() {
        val malpathString = "EN/MAL/SOM/LIGGER/HER"
        val malPath = templateRootTemp.resolve(malpathString)
        Files.createDirectories(malPath)
        val templatePath = getTemplatePath(basePathTemp, malpathString)
        Files.write(templatePath, "maldata".toByteArray())


        // expect
        val (_, content) = malService.getTemplate(malpathString)

        // when
        Assertions.assertThat(content).isEqualTo("maldata")
    }

    @Test
    fun skalReturnereIllegalArgumentExceptionVedLagreMalUtenPayloadUtenMarkdown() {
        assertThatThrownBy { malService.saveTemplate("tomPayload", "{}") }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessage("Kan ikke hente markdown for payload={}")
    }

    @Test
    @Throws(IOException::class)
    fun skalLagreMal() {

        // expect
        malService.saveTemplate(MALNAVN, GYLDIG_PAYLOAD)
        val malData = Files.readString(getTemplatePath(basePathTemp, MALNAVN))
        Assertions.assertThat(malData).isEqualTo(MARKDOWN)
    }

    @Test
    @Throws(IOException::class)
    fun skal_compile_inline_with_a_footer() {
        val tmp = malService.compileInLineTemplate("this would be a header\n{{> footer }}")
        val node: JsonNode = JsonNodeFactory.instance.arrayNode()
        val compiledStr = tmp!!.apply(node)
        Assertions.assertThat(compiledStr).endsWith(FOOTER_CONTENT)
    }

    @Test
    @Throws(IOException::class)
    fun skal_overskrive_alt_ved_lagring_av_nytt_slankere_innhold() {
        skalLagreMal()
        malService.saveTemplate(MALNAVN, lagTestPayload("\"" + MARKDOWN.substring(3) + "\"", MERGE_FIELDS))
        val malData = Files.readString(getTemplatePath(basePathTemp, MALNAVN))
        Assertions.assertThat(malData).isEqualTo(MARKDOWN.substring(3))
    }

    @Test
    @Throws(IOException::class)
    fun skalHenteHtml() {
        val malNavn = "html"
        FileUtils.writeStringToFile(
            getTemplateSchemaPath(basePathTemp, malNavn).toFile(),
            TOM_JSON,
            StandardCharsets.UTF_8
        )

        // expect
        malService.saveTemplate(malNavn, GYLDIG_PAYLOAD)
        val html = malService.createHtml(malNavn, "{\"name\":\"Peter\"}")
        Assertions.assertThat(html).contains("<h1>Hei, Peter</h1>")
    }

    @Test
    @Throws(IOException::class)
    fun skalHentePdf() {
        val malNavn = "pdf"
        FileUtils.writeStringToFile(
            getTemplateSchemaPath(basePathTemp, malNavn).toFile(),
            TOM_JSON,
            StandardCharsets.UTF_8
        )

        // expect
        malService.saveTemplate(malNavn, GYLDIG_PAYLOAD)
        val pdf = malService.createPdf(malNavn, GYLDIG_PAYLOAD)
        Assertions.assertThat(pdf[1]).isEqualTo(0x50.toByte()) //P
        Assertions.assertThat(pdf[2]).isEqualTo(0x44.toByte()) //D
        Assertions.assertThat(pdf[3]).isEqualTo(0x46.toByte()) //F
    }

    @Test
    @Throws(IOException::class)
    fun skalHentePdfFraVariation() {
        val malNavn = "pdf"
        val templateVariation = "template_NN"
        FileUtils.writeStringToFile(
            getTemplateSchemaPath(basePathTemp, malNavn).toFile(),
            TOM_JSON,
            StandardCharsets.UTF_8
        )

        // expect
        malService.saveTemplate(malNavn, GYLDIG_PAYLOAD, templateVariation)
        val pdf = malService.createPdf(malNavn, GYLDIG_PAYLOAD, templateVariation)
        Assertions.assertThat(pdf[1]).isEqualTo(0x50.toByte()) //P
        Assertions.assertThat(pdf[2]).isEqualTo(0x44.toByte()) //D
        Assertions.assertThat(pdf[3]).isEqualTo(0x46.toByte()) //F
    }

    @Test
    fun skalCompileNorwegianDatetimeHelper() {
        val jsonNode: JsonNode = JsonNodeFactory.instance.objectNode().apply { put("timestamp", "2023-08-10T15:54:35") }
        val context = Context.newBuilder(jsonNode).resolver(JsonNodeValueResolver.INSTANCE).build()
        val tmpl = malService.compileInLineTemplate("{{norwegian-datetime timestamp}}")
        val compiledStr = tmpl!!.apply(context)
        Assertions.assertThat(compiledStr).isEqualTo("10.08.2023 15:54")
    }

    companion object {
        private const val MALNAVN = "lagretMal"
        private const val MARKDOWN = "# Hei, {{name}}"
        private const val MARKDOWN_CONTENT = "\"" + MARKDOWN + "\""
        private const val MERGE_FIELDS = "{\"name\": \"Peter\"}"
        private val GYLDIG_PAYLOAD = lagTestPayload(MARKDOWN_CONTENT, MERGE_FIELDS)
        private const val FOOTER_CONTENT = "this would be the footer."
        private fun lagTestPayload(markdownContent: String, interleavingFields: String): String {
            return "{\"markdownContent\": " + markdownContent +
                    ", \"interleavingFields\": " + interleavingFields +
                    ", \"useTestSet\": false}"
        }
        private const val TOM_JSON = "{}"
    }
}